### PR TITLE
Initial working version of using glamor core

### PR DESCRIPTION
**Pull request by tkh44**
**What**:
Use glamor's core algorithm instead of cxs'

**Why**:
CSS object support is much cleaner. This will lead to fixes related to #128, #104 eventually as well.

**How**:
I ported glamor's core to work in our setup

**Checklist**:

- [ ] Documentation
- [ ] Tests
- [ ] Code complete

I did this on @threepointone's advice.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -112,7 +113,7 @@
   "bundlesize": [
     {
       "path": "./dist/DO-NOT-USE.min.js",
-      "threshold": "3 Kb"
+      "threshold": "6 Kb"
```
Comment (thangngoc89) [inline package.json]: Wow. This looks scarryyyyyyyyyyyy

Comment (tkh44) [inline package.json]: The react bundle went from 4k to 5k

Comment (thangngoc89) [inline package.json]: Well, that's acceptable to me. Probably need a seperate preact bundle again since preact don't need to filter props.

Comment (tkh44) [inline package.json]: Keep in mind this is my first pass at getting glamor's core algo working here. I keep finding code from features that I don't need and removing it.

File: src/glamor/CSSPropertyOperations/CSSProperty.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+/**
+ * Support style names that may come passed in prefixed by adding permutations
+ * of vendor prefixes.
+ */
+let prefixes = ['Webkit', 'ms', 'Moz', 'O']
+
+// Using Object.keys here, or else the vanilla for-in loop makes IE8 go into an
+// infinite loop, because it iterates over the newly added props too.
+Object.keys(isUnitlessNumber).forEach(function (prop) {
```
Comment (lukeed) [inline src/glamor/CSSPropertyOperations/CSSProperty.js]: Change `isUnitlessNumber` to an array of names, iterated to create the final object with standard & prefixed names.

10-30x faster & less code.

Also, capitalize `prefixes` already. It's only used here (`prefixKey`) and `ms` is the only lowercase-leading entry.

File: src/glamor/CSSPropertyOperations/dangerousStyleValue.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  if (
+    isNonNumeric ||
+    value === 0 ||
+    (isUnitlessNumber.hasOwnProperty(name) && isUnitlessNumber[name])
+  ) {
+    return '' + value // cast to string
+  }
+
+  if (typeof value === 'string') {
+    if (process.env.NODE_ENV !== 'production') {
```
Comment (lukeed) [inline src/glamor/CSSPropertyOperations/dangerousStyleValue.js]: Cache `process.env.NODE_ENV` to a top-level variable. Accessing this is ~1ms (credit Inferno).

File: src/glamor/CSSPropertyOperations/index.js [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+  let warnedStyleValues = {}
+  let warnedForNaNValue = false
+
+  let warnHyphenatedStyleName = function (name, owner) {
+    if (warnedStyleNames.hasOwnProperty(name) && warnedStyleNames[name]) {
+      return
+    }
+
+    warnedStyleNames[name] = true
+    process.env.NODE_ENV !== 'production'
```
Comment (lukeed) [inline src/glamor/CSSPropertyOperations/index.js]: I'd move all these `!isProd ? warning(...) : void 0` to a helper function -- probably a small wrapper around existing `warning`.

**Comment by nishp1** [comment]
> This will lead to fixes related to #128, #104 eventually as well.

#128 also applies to tagged template literals as well. Would glamor help with that too? Just want to make sure, we are solving both variants if we are going to bring extra bytes.

**Comment by nishp1** [comment]
Would also recommend changing for loops to use `@arr/forEach`.

File: src/glamor/CSSPropertyOperations/CSSProperty.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+  outline: {
+    outlineWidth: true,
+    outlineStyle: true,
+    outlineColor: true
+  }
+}
+
+const CSSProperty = {
+  isUnitlessNumber: isUnitlessNumber,
+  shorthandPropertyExpansions: shorthandPropertyExpansions
```
Comment (nishp1) [inline src/glamor/CSSPropertyOperations/CSSProperty.js]: This doesn't seem to used elsewhere, so can be removed.

File: src/index.js [hunk 6]
```diff
@@ -93,94 +98,400 @@ export function hydrate (ids: string[]) {
...
 
-  return className
+// define some constants
+
+const isDev = process.env.NODE_ENV === 'development' || !process.env.NODE_ENV
+const isTest = process.env.NODE_ENV === 'test'
+
+// takes a string, converts to lowercase, strips out nonalphanumeric.
+function simple (str) {
+  return str.toLowerCase().replace(/[^a-z0-9]/g, '')
```
Comment (nishp1) [inline src/index.js]: Regex should be moved out of func as a const.

File: src/index.js [hunk 7]
```diff
@@ -93,94 +98,400 @@ export function hydrate (ids: string[]) {
...
-function deconstruct (selector, styles, media) {
-  const decs = []
-  const rules = []
+// of shape { 'data-css-<id>': '' }
+export function isLikeRule (rule: GlamorRule) {
+  let keys = Object.keys(rule).filter(x => x !== 'toString')
+  if (keys.length !== 1) {
+    return false
+  }
+  return !!/css\-([a-zA-Z0-9]+)/.exec(keys[0])
```
Comment (nishp1) [inline src/index.js]: Same as https://github.com/tkh44/emotion/pull/138/files#r127111313

File: src/index.js [hunk 8]
```diff
@@ -93,94 +98,400 @@ export function hydrate (ids: string[]) {
...
+}
 
-  for (let key in styles) {
-    const value = styles[key]
-    const type = typeof value
+// extracts id from a { 'css-<id>': ''} like object
+export function idFor (rule: GlamorRule) {
+  let keys = Object.keys(rule).filter(x => x !== 'toString')
+  if (keys.length !== 1) throw new Error('not a rule')
+  let regex = /css\-([a-zA-Z0-9]+)/
```
Comment (nishp1) [inline src/index.js]: https://github.com/tkh44/emotion/pull/138/files#r127111313

File: src/index.js [hunk 9]
```diff
@@ -35,17 +38,66 @@ export function flush () {
...
+  return rule
+}
+
+// The idea on how to merge object class names come from glamorous
+// 💄
+// https://github.com/paypal/glamorous/blob/master/src/get-glamor-classname.js
+function getGlamorStylesFromClassName (className) {
+  const id = className.trim().slice('css-obj-'.length)
+  if (sheet.registered[id]) {
+    return sheet.registered[id].style
```
Comment (mitchellhamilton) [inline src/index.js]: Should this be `registered` or should the other references to `registered` be `sheet.registered` or am I totally wrong?

File: src/glamor/CSSPropertyOperations/CSSProperty.js [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+export default CSSProperty
```
Comment (lukeed) [inline src/glamor/CSSPropertyOperations/CSSProperty.js]: At least in context of this PR, `isUnitlessNumber` is the only thing used from this file. 

Regardless, should use `export default { isUnitlessNumber }` as opposed to exporting a created Object. This way you can actually import what you need/use & shake away the rest. This alone would probably drop a 1KB and make @thangngoc89 happy lol.

Comment (thangngoc89) [inline src/glamor/CSSPropertyOperations/CSSProperty.js]: @lukeed lol. I'm sensitive about size :D . Everything must be small and fast. 

P/s: Actually, my target platform is 512MB Android phone so you know :D

Comment (tkh44) [inline src/glamor/CSSPropertyOperations/CSSProperty.js]: I'm going to use the version from cxs so I can just avoid using facebook code.

https://github.com/tkh44/emotion/blob/master/src/index.js#L157-L186

Comment (thangngoc89) [inline src/glamor/CSSPropertyOperations/CSSProperty.js]: @tkh44 that sounds better. But perhaps just an array instead? `indexOf` is fast

Comment (lukeed) [inline src/glamor/CSSPropertyOperations/CSSProperty.js]: I checked that snippet earlier. It's faster to do `obj[key] !== undefined` than it is to do `arr.indexOf(key) !== -1`

**Pull request merged**